rts/motoko-rts/src/stabilization/layout/stable_array.rs

```diff
@@ -0,0 +1,130 @@
+use crate::{
+    memory::{alloc_array, Memory},
+    stabilization::{
+        deserialization::stable_memory_access::StableMemoryAccess,
+        graph_copy::GraphCopy,
+        serialization::{
+            stable_memory_stream::{ScanStream, StableMemoryStream, WriteStream},
+            ArraySlice, SerializationContext,
+        },
+    },
+    types::{size_of, Array, Value, TAG_ARRAY},
+};
+
+use super::{Serializer, StableToSpace, StableValue, StaticScanner};
+
+#[repr(C)]
+pub struct StableArray {
+    array_length: u64,
+    // Dynamically sized body with `array_length` elements, each of `StableValue`.
+}
+
+impl StaticScanner<StableValue> for StableArray {}
+
+impl Serializer<Array> for StableArray {
+    unsafe fn serialize_static_part(
+        _stable_memory: &mut StableMemoryStream,
+        array: *mut Array,
+    ) -> Self {
+        StableArray {
+            array_length: array.len() as u64,
+        }
+    }
+
+    unsafe fn serialize_dynamic_part(
+        stable_memory: &mut StableMemoryStream,
+        main_array: *mut Array,
+    ) {
+        for index in 0..main_array.len() {
+            let main_element = main_array.get(index);
+            let stable_element = StableValue::serialize(main_element);
+            stable_memory.write(&stable_element);
+        }
+    }
+
+    fn scan_serialized_dynamic<
+        'a,
+        M,
+        F: Fn(&mut SerializationContext<'a, M>, StableValue) -> StableValue,
+    >(
+        &self,
+        context: &mut SerializationContext<'a, M>,
+        translate: &F,
+    ) {
+        debug_assert!(!context.serialization.pending_array_scanning());
+        let mut index = 0;
+        self.sliced_array_scan(context, translate, &mut index);
+        if index < self.array_length {
+            let slice = ArraySlice::new(index, self.array_length);
+            context.serialization.set_array_slice(slice);
+        }
+    }
+
+    unsafe fn allocate_deserialized<M: Memory>(&self, main_memory: &mut M) -> Value {
+        let array_length = self.array_length as usize;
+        alloc_array(main_memory, array_length)
+    }
+
+    unsafe fn deserialize_static_part(&self, target_array: *mut Array) {
+        debug_assert_eq!((*target_array).header.tag, TAG_ARRAY);
+        debug_assert_eq!((*target_array).len as u64, self.array_length);
+    }
+
+    unsafe fn deserialize_dynamic_part<M: Memory>(
+        &self,
+        _main_memory: &mut M,
+        stable_memory: &StableMemoryAccess,
+        stable_object: StableValue,
+        target_array: *mut Array,
+    ) {
+        let stable_address = stable_object.payload_address();
+        let mut element_address =
+            stable_address + size_of::<StableArray>().to_bytes().as_usize() as u64;
+        for index in 0..(*target_array).len {
+            let element = stable_memory.read::<StableValue>(element_address);
```

Comment:
Ditto here:this loop has to complete in one increment --- is it possible that a large array will cause deserialization to fail? Should we maybe worry about making this incremental too?

Reply:
I will check, but I believe, even for very large arrays, it should be okay. The critical loop is the scanning which is incremental.

Reply:
For an array of length 100_000_000, I measured 6 billion instructions for deserialization. Probably, this is fine for the 40 billion instruction update message in which the increment runs.